internal/controller/dnsrecord_controller.go

```diff
@@ -133,36 +128,83 @@ func (r *DNSRecordReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 		setDNSRecordCondition(dnsRecord, string(v1alpha1.ConditionTypeReady), metav1.ConditionFalse, reason, message)
-		return r.updateStatus(ctx, previous, dnsRecord, noRequeueDuration, err)
+		return r.updateStatus(ctx, previous, dnsRecord, hadChanges, err)
 	}
-	// success
-	dnsRecord.Status.ObservedGeneration = dnsRecord.Generation
-	dnsRecord.Status.Endpoints = dnsRecord.Spec.Endpoints
 
-	if err := r.ReconcileHealthChecks(ctx, dnsRecord); err != nil {
+	if err = r.ReconcileHealthChecks(ctx, dnsRecord); err != nil {
 		return ctrl.Result{}, err
 	}
 
-	return r.updateStatus(ctx, previous, dnsRecord, requeueAfter, nil)
+	return r.updateStatus(ctx, previous, dnsRecord, hadChanges, nil)
 }
 
-func (r *DNSRecordReconciler) updateStatus(ctx context.Context, previous, current *v1alpha1.DNSRecord, requeueAfter time.Duration, specErr error) (reconcile.Result, error) {
+func (r *DNSRecordReconciler) updateStatus(ctx context.Context, previous, current *v1alpha1.DNSRecord, hadChanges bool, specErr error) (reconcile.Result, error) {
+	var requeueTime time.Duration
+	logger := log.FromContext(ctx)
+
+	// short loop. We don't publish anything so not changing status
+	if prematurely, requeueIn := recordReceivedPrematurely(current); prematurely {
+		return reconcile.Result{RequeueAfter: requeueIn}, nil
+	}
+
+	// failure
 	if specErr != nil {
-		current.Status.WriteCounter = previous.Status.WriteCounter
-		current.Status.ValidFor = previous.Status.ValidFor
-		current.Status.QueuedAt = previous.Status.QueuedAt
-		current.Status.QueuedFor = previous.Status.QueuedFor
+		var updateError error
+		if !equality.Semantic.DeepEqual(previous.Status, current.Status) {
+			if updateError = r.Status().Update(ctx, current); updateError != nil && apierrors.IsConflict(updateError) {
+				return ctrl.Result{Requeue: true}, nil
+			}
+		}
+		return ctrl.Result{Requeue: true}, updateError
+	}
+
+	// success
+	if hadChanges {
+		// generation has not changed but there are changes.
+		// implies that they were overridden - bump write counter
+		if !generationChanged(current) {
+			current.Status.WriteCounter++
+			wrtiteCounter.WithLabelValues(current.Name, current.Namespace).Inc()
+			logger.V(1).Info("Changes needed on the same generation of record")
+		}
+		requeueTime = validationRequeueTime
+		setDNSRecordCondition(current, string(v1alpha1.ConditionTypeReady), metav1.ConditionFalse, "AwaitingValidation", "Awaiting validation")
+	} else {
+		logger.Info("All records are already up to date")
+		// reset the valid for from randomized value to a fixed value once validation succeeds
+		if !meta.IsStatusConditionTrue(current.Status.Conditions, string(v1alpha1.ConditionTypeReady)) {
+			requeueTime = exponentialRequeueTime(validationRequeueTime.String())
+		} else {
+			// uses current.Status.ValidFor as the last requeue duration. Double it.
+			requeueTime = exponentialRequeueTime(current.Status.ValidFor)
```

Comment:
this still reads to me like:
```
if it wasn't previously ready {
   requeue for randomized time
} else {
   requeue for double the last requeue time
}
```

I'm not seeing how this avoids doubling the randomized number.

Reply:
used the wrong var while moving/renaming things.
Now it should be fine